gazelle_haskell_modules/rule_generation.go

```diff
@@ -265,21 +267,60 @@ func concatRuleInfos(xs [][]*RuleInfo) []*RuleInfo {
 	return ys
 }
 
-// Collects the source files referenced in the given expression
-func srcsFromRule(pkgRoot string, expr build.Expr) ([]string, error) {
-	srcs, err := getSources(expr)
-	if err != nil {
-		return nil, err
+// Find either
+// 0. The files listed in the srcs attribute of a rule
+// 1. All the files (recursively) in directories specified in a "gazelle_haskell_modules:srcs:"
+// If the "srcs" attribute is present, 1. is ignored.
+func getSrcs(pkgRoot string, r *rule.Rule) ([]string, error) {
+	hasSrcs := false
+	for _, attrKey := range r.AttrKeys() {
+		if attrKey == "srcs" {
+			hasSrcs = true
+			break
+		}
 	}
 
-	xs := make([]string, len(srcs))
-	i := 0
-	for f, _ := range srcs {
-		xs[i] = path.Join(pkgRoot, f)
-		i++
-	}
+	if hasSrcs {
+		srcs, err := getSources(r.Attr("srcs"))
 
-	return xs, nil
+		if err != nil {
+			return nil, err
+		}
+
+		xs := make([]string, len(srcs))
+		i := 0
+		for f, _ := range srcs {
+			xs[i] = path.Join(pkgRoot, f)
+			i++
+		}
+
+		return xs, nil
+	} else {
+		srcDirs, err := getSrcDirsFromRuleDirective(r)
+		if err != nil {
+			log.Fatal(err)
```

Comment:
For me, using `log.Fatal` is a matter of whether we have in the context all the information that is needed to make an intelligible error message. If you kill gazelle here, will the user have it easy understanding which build file and which rule in it is causing the error. Consider that this can be used on large repos.

Reply:
I tried to wrap it appropriately, thanks.